### DISCUSSION TITLE
Unauthorized page shown when user is following a link to the wiki and isn't authenticated

### DISCUSSION THREAD

**Discussion by PaulD987** [requarks/wiki#3744] [Error / Bug Report]
Our wiki is set up to not allow any anonymous access etc... eg Guest group has no permissions. If a user visits the homepage, then they are redirected to the login page, which is all as would be expected. If however, the user visits a link to a page in the wiki, and they aren't already authenticated with the wiki, they are presented with an "Unauthorized - You cannot view this page." message. Whilst there is a "login as" button on that page, it still seems to confuse people and I get messages saying that they aren't authorised... whereas all they actually need to do it sign in. Would be nice if it would always redirect to the login page if they weren't authenticated. Maybe an option to enable this behaviour if we don't want to change the default behaviour for everyone.

**Comment by PaulD987** [comment]
I'd be happy to have a go implementing this, but would prefer an indication before hand that it is a fix that would be generally acceptable, and whether it should be an optional config thing. Maybe on Security page,  "Always redirect to login page if not authenticated" option.

**Comment by PaulD987** [comment]
in /server/controllers/common.js the route `router.get('/*'` contains
```
        if (pageArgs.path === 'home' && req.user.id === 2) {
          return res.redirect('/login')
        }
```
This means that it will only redirect straight to the login page if the person trying to view the page that they aren't authorised to see is "guest" and the page is the home page. I'm not sure why it should matter whether it is the home page or not. Surely it would be more logical to just redirect to login all the time the "guest" user, someone who hasn't authenticated, hits a page that they aren't authorised to view.

**Comment by PaulD987** [comment]
@NGPixel sorry to trouble you, but could I get your view on this please? I have coded this, including configuration option on Security tab to decide whether the user wants to carry with current behaviour, or switch to only having unauthorised message if not authenticated.

![image](https://user-images.githubusercontent.com/73899671/113008410-e9361d80-916e-11eb-8530-72a8ea03d8e1.png)

Test then becomes:
```
if ((pageArgs.path === 'home' || WIKI.config.auth.bypassUnauthorized) && req.user.id === 2) {
          return res.redirect('/login')
        }
```

Only bit that I haven't managed to get working in my testing is using locale files to load the strings properly for the security page. If I submit a PR, would you prefer to have the option to configure behaviour? or just always redirect to login?  If former, how do the new strings get added?

```
    "admin:security.bypassUnauthorized": "Bypass Unauthorized Screen",
    "admin:security.bypassUnauthorizedHint": "Should the user be redirected automatically to the login screen if they are not authenticated and attempt to access a page not accessible to Guest"
```

**Reply by NGPixel** [reply]
It should definitely be an option. You're assuming that everyone landing on a page should be able to login, which is not always the case and shouldn't be the default.

**Reply by PaulD987** [reply]
Thanks, yes, that would certainly be a fairly standard use case. For our wiki, all pages are protected, and therefore the only option really for anyone is to attempt to log in so the "unauthorised" page is just an annoyance, hence why I would like to avoid it for us. OK if I put in a PR? and if so, how do the strings get added?

**Reply by NGPixel** [reply]
Sure, submit a PR. For the locale strings, simply post them in the PR ticket and I'll add them in Lokalise.

**Reply by PaulD987** [reply]
Thanks, I appreciate your help!
